build_compilers/lib/src/workers.dart

```diff
@@ -9,76 +9,134 @@ import 'dart:math' show min;
+      maxWorkers: _maxWorkersPerTask,
+    );
+  }
+
+  /// Executes the following command [args] against the worker.
+  ///
+  /// Returns a [WorkResponse] object from `package:bazel_worker`.
+  Future<WorkResponse> execute(List<String> args) async {
+    return _driver.doWork(new WorkRequest()..arguments.addAll(args));
+  }
+
+  /// Completes when the worker is considered complete and shut down.
+  ///
+  /// A worker that has not started is always considered "done".
+  ///
+  /// TODO: Consider different naming, i.e. is/isNotActive(?).
+  Future<Null> get isDone => _onDone?.future ?? new Future.value();
+
+  /// Resource for accessing the spawned process.
+  ///
+  /// Can be used with [BuildStep.fetchResource].
+  Resource<BatchWorker> get resource {
+    return _resource ??= new Resource<BatchWorker>(
+      () {
+        _createDriver();
+        return this;
+      },
+      beforeExit: () async {
+        await _driver?.terminateWorkers();
+        _driver = null;
+        _onDone?.complete();
+        _onDone = null;
+      },
+    );
+  }
+
+  /// Starts a persistent worker in [workingDirectory].
+  @protected
+  Future<Process> startProcess(String workingDirectory);
 }
 
-BazelWorkerDriver __analyzerDriver;
-
-/// Resource for fetching the current [BazelWorkerDriver] for dartanalyzer.
-final analyzerDriverResource = new Resource<BazelWorkerDriver>(
-    () => _analyzerDriver, beforeExit: () async {
-  await _analyzerDriver?.terminateWorkers();
-  _analyzerWorkersAreDoneCompleter.complete();
-  _analyzerWorkersAreDoneCompleter = null;
-  __analyzerDriver = null;
-});
-
-/// Manages a shared set of persistent dartdevc workers.
-BazelWorkerDriver get _dartdevcDriver {
-  _dartdevcWorkersAreDoneCompleter ??= new Completer<Null>();
-  return __dartdevcDriver ??= new BazelWorkerDriver(
-      () => Process.start(p.join(sdkDir, 'bin', 'dartdevc$_scriptExtension'),
-          ['--persistent_worker'],
-          workingDirectory: scratchSpace.tempDir.path),
-      maxWorkers: _maxWorkersPerTask);
+final BatchWorker analyzerWorker = new _AnalyzerWorker();
+
+class _AnalyzerWorker extends BatchWorker {
+  static final _defaultPath = _sdkBin('dartanalyzer');
+
+  @override
+  Future<Process> startProcess(String workingDirectory) {
+    return Process.start(
+      _defaultPath,
+      const [
+        '--build-mode',
+        '--persistent_worker',
+      ],
+      workingDirectory: workingDirectory,
+    );
+  }
 }
 
-BazelWorkerDriver __dartdevcDriver;
+final BatchWorker dartdevcWorker = new _DartDevCWorker();
 
-/// Resource for fetching the current [BazelWorkerDriver] for dartdevc.
-final dartdevcDriverResource = new Resource<BazelWorkerDriver>(
-    () => _dartdevcDriver, beforeExit: () async {
-  await _dartdevcDriver?.terminateWorkers();
-  _dartdevcWorkersAreDoneCompleter.complete();
-  _dartdevcWorkersAreDoneCompleter = null;
-  __dartdevcDriver = null;
-});
+class _DartDevCWorker extends BatchWorker {
```

Comment:
Could we try a pattern without `extends`? If we don't think we'll need to `@override` other methods in the class I really like when we're precise about what behavior can possibly be changed.

```dart
typedef Future<Process> StartProcess(String workingDirectory);

class BatchWorker {
  final StartProcess _startProcess;
  ...
  BatchWorker(this._startProcess);
}

final dartDevcWorker = new BatchWorker((workingDirectory) => Process.start(
    _defaultPath, const ['--persistent_worker'],
    workingDirectory: workignDirectory));
```